[draft] Reject unencrypted incoming mail

hpk42 opened the pull request:
This passes tests but it's not complete because merging would break existing cleartext-usages of newsletter-subscriptions, and more, importantly, of throwaway login-emails.

hpk42 commented on cmdeploy/src/cmdeploy/tests/plugin.py:
```diff
@@ -62,7 +62,7 @@ def sshdomain(maildomain):
     domain = os.environ.get("CHATMAIL_DOMAIN2")
     if not domain:
-        pytest.skip("set CHATMAIL_DOMAIN2 to a ssh-reachable chatmail instance")
+        pytest.skip("set CHATMAIL_DOMAIN2 to second chatmail server (FQDN form)")
```
FQDN does not imply there is a trailing dot, does it?  It's not sufficient to use an ssh-name because the test attempts to create an e-mail address

hpk42 commented on chatmaild/src/chatmaild/user.py:
```diff
@@ -13,6 +13,7 @@ def __init__(self, maildir, addr, password_path, uid, gid):
         self.maildir = maildir
         self.addr = addr
         self.password_path = password_path
+        self.enforce_E2EE_path = maildir.joinpath("enforceE2EE")
```
i wanted to have a short name.  E2EE  in the context of chatmail is defined as OpenPGP. don't think there is ambiguity.  maybe `enforceE2EEincoming` -- don't want to introduce Camelcase.

adbenitez commented on chatmaild/src/chatmaild/filtermail.py:
```diff
@@ -230,6 +250,61 @@ def check_DATA(self, envelope):
         for recipient in envelope.rcpt_tos:
             if recipient_matches_passthrough(recipient, passthrough_recipients):
                 continue
+            p = self.config.mailboxes_dir.joinpath(recipient, "inclear")
```
what is this "inclear"??? sorry if I overlooked the explanation somewhere :thinking:

hpk42 replied:
no, it's a very good point! this is left-over and needs to change, but i first need to figure out why no test complained.

hpk42 replied:
fixed in https://github.com/chatmail/server/pull/538/commits/e74e942ed2131fc5fcf1ab2f2f3e4f3e4d6ef843

adbenitez commented on cmdeploy/src/cmdeploy/service/filtermail-incoming.service.f:
```diff
@@ -0,0 +1,12 @@
+Description=Incoming Chatmail Postfix before queue filter 
+
+[Service]
+ExecStart={execpath} {config_path} incoming
```
this could have been a single service, but I guess it is interesting to see which one causes more memory usage, OTOH what is likely to happen is that both will use more memory usage than having a single service

adbenitez commented:
it seems to be missing some explanation/docs about how to use the `enforceE2EE` file hack and there seems to also be an "inclear" file hack?

hpk42 commented:
> it seems to be missing some explanation/docs about how to use the `enforceE2EE` file to allow unencrypted etc. and there seems to also be an undocumented "inclear" file hack?

added a section in the README https://github.com/chatmail/server/pull/538/commits/4b0d7a0fa282f72341fe7466d244f96249ae710e  -- (and removed the wrong "inclear" left-overs in the previous commit)

missytake commented on chatmaild/src/chatmaild/filtermail.py:
```diff
@@ -214,24 +236,75 @@ def check_DATA(self, envelope):
+                return
 
         for recipient in envelope.rcpt_tos:
-            if recipient_matches_passthrough(recipient, passthrough_recipients):
+            print(f"filtering recipient: {recipient}")
+            if self.config.get_user(recipient).is_incoming_cleartext_ok():
                 continue
```
Does envelope.rcpt_tos only contain recipients on the own chatmail server? Others wouldn't be in our vmail directory and we couldn't create a user object for them...

hpk42 replied:
i think postfix's smtpd on port 25, before it passes the mail to the filtermail-incoming service, checks whether the recipient mailboxes exist: when i tried to send a cleartext message to a not-existing address, it was rejected and filtermail never saw the mail. but i'll modify this to be on the safe side.

missytake reviewed (approved):
Looks reasonable, all in all :)

hpk42 commented on chatmaild/src/chatmaild/filtermail.py:
```diff
@@ -214,24 +236,76 @@ def check_DATA(self, envelope):
 
         for recipient in envelope.rcpt_tos:
-            if recipient_matches_passthrough(recipient, passthrough_recipients):
+            print(f"filtering recipient: {recipient}")
```
yes but i removed it.  it's too verbose.

hpk42 commented on chatmaild/src/chatmaild/filtermail.py:
```diff
@@ -214,24 +236,76 @@ def check_DATA(self, envelope):
+        # chatmail core will typically not display them as normal messages
+        if message.get("auto-submitted"):
+            _, from_addr = parseaddr(message.get("from").strip())
+            if from_addr.lower().startswith("mailer-daemon@"):
```
checking now for it.
